azure-application-insights-spring-boot-starter/src/main/java/com/microsoft/applicationinsights/boot/ApplicationInsightsModuleConfiguration.java

```diff
@@ -0,0 +1,155 @@
+    @Bean
+    public DeviceInfoContextInitializer deviceInfoContextInitializer() {
+        return new DeviceInfoContextInitializer();
+    }
+
+    /**
+     * Bean for SpringBootTelemetryInitializer
+     * @return instance of {@link SpringBootTelemetryInitializer}
+     */
+    @Bean
+    public SpringBootTelemetryInitializer springBootTelemetryInitializer() {
+        return new SpringBootTelemetryInitializer();
+    }
+
+    /**
+     * Bean for ProcessPerformanceCounterModule
+     * @return instance of {@link ProcessPerformanceCountersModule}
+     */
+    //FIXME: This should be conditional on operating System. However, current architecture of ProcessBuiltInPerformanceCountersFactory
+    //FIXME: does not separate this concerns therefore cannot condition as of now.
+    @Bean
+    @DependsOn("performanceCounterContainer")
+    @ConditionalOnProperty(value = "azure.application-insights.default-modules.ProcessPerformanceCountersModule.enabled", havingValue = "true", matchIfMissing = true)
+    public ProcessPerformanceCountersModule processPerformanceCountersModule() {
+        try {
+            return new ProcessPerformanceCountersModule();
+        }
+        catch (Exception e) {
+            throw new IllegalStateException("Could not initialize Windows performance counters module, " +
+                    "please set property 'azure.application-insights.default-modules.ProcessPerformanceCountersModule.enabled=false' to avoid this error message.", e);
+        }
+    }
+
+    /**
+     * Bean for JvmPerformanceCounterModule
+     * @return instance of {@link JvmPerformanceCountersModule}
+     */
+    @Bean
+    @DependsOn("performanceCounterContainer")
+    @ConditionalOnProperty(value = "azure.application-insights.default.modules.JvmPerformanceCountersModule.enabled", havingValue = "true", matchIfMissing = true)
+    public JvmPerformanceCountersModule jvmPerformanceCountersModule() {
+        try {
+            return new JvmPerformanceCountersModule();
+        }
+        catch (Exception e) {
+            throw new IllegalStateException("Could not initialize Jvm Performance Counters module "
+                + "please set the property 'azure.application-insights.default.modules.JvmPerformanceCountersModule.enabled=false' to "
+                + "avoid this error message", e);
+        }
+    }
+
+    /**
+     * Bean for HeartBeatModule. This also sets the properties for HeartBeatModule
+     * @param environment
+     * @return initialized instance with user specified properties of {@link HeartBeatModule}
+     */
+    @Bean
+    @ConditionalOnProperty(value = "azure.application-insights.default.modules.HeartBeat.enabled", havingValue = "true", matchIfMissing = true)
+    public HeartBeatModule heartBeatModule(Environment environment) {
+        try {
+            HeartBeatModule heartBeatModule = new HeartBeatModule();
+            HeartbeatDefaultPayload.addDefaultPayLoadProvider(new SpringBootHeartBeatProvider(environment));
```

Comment:
If it's expected from users to be able to provide own `HeartBeatProvider` implementation it's worth to create bean:
```
@Bean
@ConditionalOnMissingBean
public HeartBeatPayloadProviderInterface heartBeatPayloadProviderInterface(Environment environment) {
    return new SpringBootHeartBeatProvider(environment);
}
```
and inject `HeartBeatPayloadProviderInterface` into `HeartBeatModule`. It's important to have interface declaration instead of implementation - `SpringBootHeartBeatProvider` because only this way users are able to override the bean.

Reply:
In all cases we would like to have the properties of the current provider for better assessment. However it makes sense to do this more elegantly using bean. I would refactor this.

Reply:
@gavlyukovskiy this is now addressed.